org.eclipse.lsp4e/src/org/eclipse/lsp4e/LSPEclipseUtils.java

```diff
@@ -217,10 +219,18 @@ public static boolean isOffsetInRange(int offset, Range range, IDocument documen
 		}
 	}
 
-	public static CompletionParams toCompletionParams(URI fileUri, int offset, IDocument document)
+	public static CompletionParams toCompletionParams(URI fileUri, int offset, IDocument document, char[] completionTriggerChars)
 			throws BadLocationException {
 		Position start = toPosition(offset, document);
 		final var param = new CompletionParams();
+		try {
+			String triggerCharacter = document.get(offset-1, 1);
+			if (new String(completionTriggerChars).contains(triggerCharacter)) {
```

Comment:
>  but auto completion is disabled

Do you mean when it's disabled on the server oder the server does not support auto completion?
I assumed that `capabilities.getCompletionProvider()` called in `org.eclipse.lsp4e.operations.completion.LSContentAssistProcessor.computeCompletionProposals(ITextViewer, int)`  will return `null` in this case and the completion proposals would not work at all. So the action would not be invoked.

Reply:
> if the user places the cursor at one position without typing

This would not trigger the completion. It will be triggered when the user enter a character. E.g.:
When I set the cursor between the first two colons :
```cpp
std::cout << "Hello World" << std::endl;
```
and enter a new colon (which is a triggerCharacter in C) I get proposals for `std::` which could make no sense in this context, when Evaluation the whole line. But I guess this is more a server issue: How to offer valid completion proposal results.

Reply:
Maybe somehting like:
`Note: the context heuristic might render wrong completion results from the server, depending on the triggerCharacters context`

But I am not sure if this confuses more than it make things clearer.

Reply:
> > but auto completion is disabled
> 
> Do you mean when it's disabled on the server oder the server does not support auto completion? I assumed that `capabilities.getCompletionProvider()` called in `org.eclipse.lsp4e.operations.completion.LSContentAssistProcessor.computeCompletionProposals(ITextViewer, int)` will return `null` in this case and the completion proposals would not work at all. So the action would not be invoked.

I mean the case where the server supports it but the client has the feature disabled because the used does not want to have the feature. See https://github.com/eclipse-platform/eclipse.platform.ui/pull/931.

Reply:
> > if the user places the cursor at one position without typing



> > if the user places the cursor at one position without typing
> 
> This would not trigger the completion.

Yes, that is correct, but the user could then press Control-Space. Then, even though we have a trigger character at  the position, it is a manual invocation, and not triggered with a trigger character. That is the case I have in mind.

Reply:
> Maybe somehting like: `Note: the context heuristic might render wrong completion results from the server, depending on the triggerCharacters context`
> 
> But I am not sure if this confuses more than it make things clearer.

Okay, then let us not put any comment.